File: Binary_Search/src/main/java/Search_A_2D_Matrix.py
# Link to problem https://leetcode.com/problems/search-a-2d-matrix/submissions/1275436074/

from typing import List
def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

    low, high = 0, len(matrix) - 1

    while (low <= high):
        mid = (low + high) // 2
        
        if target in matrix[mid]:
            return True 

        if (matrix[mid][0] < target):
            low += 1
        else:
            high -= 1
    return False

            

        

File: Binary_Search/src/main/java/Search_A_2D_matrix.java
public class Search_A_2D_matrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        int low = 0;
        int high = matrix.length - 1;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (search(matrix[mid], target)) {
                return true;
            }

            if (matrix[mid][0] < target) {
                low++;
            } else {
                high--;
            }
        }
        return false;
    }

    public boolean search(int[] list, int target) {
        for (int i : list) {
            if (i == target){
                return true;
            }
        }
        return false;
    }
}
